roc/types.go

```diff
@@ -0,0 +1,91 @@
+// WARNING: This file has automatically been generated on Sun, 13 Oct 2019 20:21:32 CEST.
+// Code generated by https://git.io/c-for-go. DO NOT EDIT.
+
+package roc
+
+/*
+#cgo linux CFLAGS: -I"../include"
+#cgo linux LDFLAGS: -lroc
+#include "../include/roc/address.h"
+#include "../include/roc/config.h"
+#include "../include/roc/context.h"
+#include "../include/roc/frame.h"
+#include "../include/roc/log.h"
+#include "../include/roc/platform.h"
+#include "../include/roc/receiver.h"
+#include "../include/roc/sender.h"
+#include <stdlib.h>
+#include "cgo_helpers.h"
+*/
+import "C"
+import "unsafe"
+
+// Address as declared in roc/address.h:59
+type Address struct {
+	ref94efe9d7    *C.roc_address
+	allocs94efe9d7 interface{}
+}
+
+// ContextConfig as declared in roc/config.h:147
+type ContextConfig struct {
+	maxPacketSize  uint32
+	maxFrameSize   uint32
+	ref6685576b    *C.roc_context_config
+	allocs6685576b interface{}
+}
+
+// SenderConfig as declared in roc/config.h:234
+type SenderConfig struct {
+	frameSampleRate       uint32
+	frameChannels         ChannelSet
+	frameEncoding         FrameEncoding
+	packetSampleRate      uint32
+	packetChannels        ChannelSet
+	packetEncoding        PacketEncoding
+	packetLength          uint64
+	packetInterleaving    uint32
+	automaticTiming       uint32
+	resamplerProfile      ResamplerProfile
+	fecCode               FecCode
+	fecBlockSourcePackets uint32
+	fecBlockRepairPackets uint32
+	refc92542a4           *C.roc_sender_config
+	allocsc92542a4        interface{}
+}
+
+// ReceiverConfig as declared in roc/config.h:316
+type ReceiverConfig struct {
+	frameSampleRate         uint32
+	frameChannels           ChannelSet
+	frameEncoding           FrameEncoding
+	automaticTiming         uint32
+	resamplerProfile        ResamplerProfile
+	targetLatency           uint64
+	maxLatencyOverrun       uint64
+	maxLatencyUnderrun      uint64
+	noPlaybackTimeout       int64
+	brokenPlaybackTimeout   int64
+	breakageDetectionWindow uint64
+	ref9cfc9e50             *C.roc_receiver_config
+	allocs9cfc9e50          interface{}
+}
+
+// Context as declared in roc/context.h:41
+type Context C.roc_context
+
+// Frame as declared in roc/frame.h:42
+type Frame struct {
+	samples       unsafe.Pointer
+	samplesSize   uint
```

Comment:
As for the `Frame` struct, I suggest to completely hide this struct from user and provide type-safe wrappers instead (https://github.com/roc-project/roc/issues/248).

Reply:
Types like this one (everything in `types.go`) are autogenerated from C header. Depending on the lib it's often required to expose everything as-is and say that this lib is "low level". As it  follows its C API that was not uder-friendly in the first place. Or, alternatively, to have a low-level generated bindings aside in private package and expose nice and clean types and methods suitable for end-user.

Example: https://github.com/xlab/pocketsphinx-go (there are nice and clean wrappers, that make calls into autogenerated Go bindings).

Reply:
Yep. I'd like to use a mixed approach. It would be convenient to export *some* generated stuff (e.g. configs) and unexport the rest and provide higher-level wrappers for it (https://github.com/roc-project/roc/pull/269#issuecomment-538951694).

Reply:
that means that we don't want to have two separate packages: one with purely generated code and one with bindings, right? if so, can we consider this convo resolved?

Reply:
> that means that we don't want to have two separate packages

Yep, I'd prefer a single package.

> if so, can we consider this convo resolved?

Nope. This comment was about hiding `Frame` type, i.e. replacing it with `frame`, and replacing `Sender.Write(*Frame)` with type-safe wrappers like `Sender.WriteAudioBytes([]byte)` and `Sender.WriteAudioFloats([]float)`, as described in https://github.com/roc-project/roc/issues/248.

We can do it in future PRs, though.

Exporting something like this to the user:

```
type Frame struct {
	Samples       unsafe.Pointer
	SamplesSize   uint
}
```

would be too low-level and unfamiliar for a Go library (because of the unsafe.Pointer).

Reply:
ok, noted :+1:

Reply:
#2